rts/motoko-rts/src/stabilization.rs

```diff
@@ -0,0 +1,82 @@
+//! Graph-copy-based stabilzation on upgrades, serializing the entire stable object graph into
+//! stable memory by using a defined long-term stable storage format.
+//!
+//! This is to support potentially radical changes of the persistent main memory layout, e.g.
+//! introducing a new GC or rearranging persistent metadata. This also relies on precise value
+//! tagging to allow more advanced changes that require value metadata, e.g. specializing arrays for
+//! small element types or even downgrading to 32-bit heap layouts (provided that the amount of live
+//! data fits into a 32-bit memory).
+//!
+//! A memory compatibility check similar to enhanced orthogonal persistence has to be performed.
+//! For this purpose, the type table of the serialized object graph is also stored in stable memory
+//! and on upgrade, compared to the new program version.
+//!
+//! A versioned stable storage format even permits future evolutions of the graph copy algorithm.
+//!  
+//! See `GraphCopyStabilization.md` for the stable format specification and the employed algorithm.
+
+pub mod deserialization;
+pub mod graph_copy;
+pub mod layout;
+pub mod serialization;
+
+#[cfg(feature = "ic")]
+pub mod ic;
+
+use core::cmp::min;
+
+use crate::{
+    constants::WASM_PAGE_SIZE,
+    rts_trap_with,
+    stable_mem::{self, ic0_stable64_write, PAGE_SIZE},
+};
+
+use self::layout::StableValue;
+
+extern "C" {
+    pub fn moc_stabilization_instruction_limit() -> u64;
+    fn ic0_performance_counter(number: u32) -> u64;
+}
+
+// Dummy value used for non-stable objects that are potentially reachable from
+// stable variable because of structural subtyping or `Any`-subtyping.
+// Must be a non-skewed value such that the GC also ignores this value.
+const DUMMY_VALUE: StableValue = StableValue::from_raw(0);
+
+fn clear_stable_memory(start: u64, length: u64) {
```

Comment:
Note to myself: This needs to be incremental when called at the end of deserialization, as it may be very large length (multiple GC) and will otherwise exceed the instruction limit.

Reply:
This is now fixed. I tested it for cases of more than 10 GB of graph-stabilized memory.